pkg/util/dynamichelper/dynamichelper.go

```diff
@@ -92,14 +126,24 @@ func (dh *dynamicHelper) Ensure(ctx context.Context, objs ...kruntime.Object) er
 	return nil
 }
 
-func (dh *dynamicHelper) ensureOne(ctx context.Context, new kruntime.Object) error {
+func (dh *DynamicHelper) ensureOne(ctx context.Context, new kruntime.Object) error {
+	// Grab all possible GVKs for object
 	gvks, _, err := scheme.Scheme.ObjectKinds(new)
 	if err != nil {
 		return err
 	}
 
 	gvk := gvks[0]
+	// ASSUMPTION: The above assumes there will only ever be one gvk for the passed-in object
+	// but there may be a case where the object is not fully formed and therefore may match
+	// multiple gvks. This is a code-path that we don't test for so it may cause unintended
+	// side-effects. Therefore we log a warning for our future selves.
+	if len(gvks) > 1 {
+		_gvk := new.GetObjectKind().GroupVersionKind()
+		dh.log.Warnf("While retrieving kinds for an object \"%s\", more that one version returned which is a codepath that's not fully handled.", _gvk.String())
+	}
```

Comment:
This was a previously undocumented assumption. I've added a warning and more information in a comment here because I believe in the future when other packages implement this function that this may be the cause of bugs. However it's not clear _what_ the correct functionality should be, which will probably be clearer when future functions make use of dynamic helper's Ensure.

Reply:
I don’t want to error out, because if existing code is hitting this case but still currently working, breaking that code in prod seems like a really dangerous idea.